File: packages/ts/src/domain/approval.ts
import type { RequestId, RunId } from "./identifiers";
import type { RiskTier, ToolCall } from "./tool";

/** Raised when a gated tool call is reached and the run suspends for a human. */
export interface ApprovalRequest {
  readonly runId: RunId;
  readonly requestId: RequestId;
  readonly toolCall: ToolCall;
  readonly risk: RiskTier;
  /** Run metadata an approver may need to decide (tenant, channel, ...). */
  readonly context: Readonly<Record<string, unknown>>;
}

export const DecisionType = {
  Approve: "approve",
  Edit: "edit",
  Reject: "reject",
} as const;
export type DecisionType = (typeof DecisionType)[keyof typeof DecisionType];

export interface ApproveDecision {
  readonly type: typeof DecisionType.Approve;
  readonly approver: string;
}

export interface EditDecision {
  readonly type: typeof DecisionType.Edit;
  readonly approver: string;
  readonly editedArgs: Readonly<Record<string, unknown>>;
}

export interface RejectDecision {
  readonly type: typeof DecisionType.Reject;
  readonly approver: string;
  readonly reason: string;
}

export type ApprovalDecision = ApproveDecision | EditDecision | RejectDecision;

File: packages/py/src/airlock/domain/approval.py
"""Approval requests raised by the gate and the decisions humans return."""

from enum import StrEnum
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field

from airlock.domain.identifiers import RequestId, RunId
from airlock.domain.tool import RiskTier, ToolCall


class ApprovalRequest(BaseModel):
    """Raised when a gated tool call is reached and the run suspends."""

    model_config = ConfigDict(frozen=True)

    run_id: RunId
    request_id: RequestId
    tool_call: ToolCall
    risk: RiskTier
    context: dict[str, object]


class DecisionType(StrEnum):
    APPROVE = "approve"
    EDIT = "edit"
    REJECT = "reject"


class ApproveDecision(BaseModel):
    model_config = ConfigDict(frozen=True)

    type: Literal[DecisionType.APPROVE] = DecisionType.APPROVE
    approver: str


class EditDecision(BaseModel):
    model_config = ConfigDict(frozen=True)

    type: Literal[DecisionType.EDIT] = DecisionType.EDIT
    approver: str
    edited_args: dict[str, object]


class RejectDecision(BaseModel):
    model_config = ConfigDict(frozen=True)

    type: Literal[DecisionType.REJECT] = DecisionType.REJECT
    approver: str
    reason: str


ApprovalDecision = Annotated[
    ApproveDecision | EditDecision | RejectDecision,
    Field(discriminator="type"),
]
